### PR TITLE
lib: implement URL for easy url handling

### PR DESCRIPTION
**Pull request by swaptr**
This PR adds support to abstract URL related functionalities to a separate struct.

### CODE REVIEW

File: reckless_github/src/lib.rs [hunk 1]
```diff
@@ -24,10 +25,13 @@ mod tests {
...
         let name = "hello";
-        let url = "https://github.com/lightningd/plugins";
-        let mut repo = Github::new(name, url);
+        let url = URL::new("https://github.com/lightningd/plugins");
+        let mut repo = Github::new(name, &url);
         let repo = repo.init().await;
         assert!(repo.is_ok());
-        assert_eq!(Path::new(&(get_dir_path_from_url(url))).exists(), true);
+        assert_eq!(
+            Path::new(&(get_dir_path_from_url(&url.url_string))).exists(),
```
Comment (vincenzopalazzo) [inline reckless_github/src/lib.rs:33]:
```suggestion
            Path::new(&(get_dir_path_from_url(&url.url_string))).exists(),
```
```suggestion
            Path::new(&url.path_string)).exists(),
```

I do not know if it is the same but we can implement the method  `get_dir_path_from_url` inside the URL?

Comment (vincenzopalazzo) [reply reckless_github/src/lib.rs:33]: maybe a good place to put a FIXME?

File: reckless_github/src/repository.rs [hunk 2]
```diff
@@ -112,13 +113,16 @@ impl Repository for Github {
         debug!(
             "INITIALIZING REPOSITORY: {} {} > {}",
             self.name,
-            self.url,
-            get_dir_path_from_url(&self.url)
+            self.url.url_string,
+            get_dir_path_from_url(&self.url.url_string)
```
Comment (vincenzopalazzo) [inline reckless_github/src/repository.rs:117]: same there

File: reckless_lib/src/url.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
+use std::fmt;
+
+use crate::utils::get_reckless_dir;
+
+/// This struct will make sure our URL's are of the
+/// correct format
+#[derive(Clone)]
+pub struct URL {
+    // the url string
```
Comment (vincenzopalazzo) [inline reckless_lib/src/url.rs:9]:
```suggestion
    /// the url string
```

in this way will be a comment

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Some minor comments, but for me the PR can be merged.

I think just add a FIXME to remember that we can improve the URL implementation

File: reckless_github/src/lib.rs (hunk 1 above)
Comment (swaptr) [reply reckless_github/src/lib.rs]: Fixed. thanks.

File: reckless_lib/src/url.rs (hunk 3 above)
Comment (swaptr) [reply reckless_lib/src/url.rs]: Thanks, just fixed.

File: reckless_github/src/repository.rs (hunk 2 above)
Comment (swaptr) [reply reckless_github/src/repository.rs]: Fixed.

**Review by vincenzopalazzo — APPROVED** [review]
Very good cleanup, thanks!

**Pull request merged**